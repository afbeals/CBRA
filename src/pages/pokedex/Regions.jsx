// Extenal
import React from 'react';
import PropTypes from 'prop-types';
import { CSSTransition } from 'react-transition-group';

// Local
import ErrorBoundary from '~Components/ErrorBoundary';
import withPokedex from '~Modules/pokedex/withPokedex';

// Constants
const classname = 'regions';
const classNames = {
  appear: 'appear',
  appearActive: 'appear__active',
  appearDone: 'appear__done',
  enter: 'enter',
  enterActive: 'enter__active',
  enterDone: 'enter__done',
  exit: 'exit',
  exitActive: 'exit__active',
  exitDone: 'exit__done',
};

// Component
const Regions = ({
  fetchDex,
  pClassname,
  getRegions,
  selectRegion,
  getSelectedRegion,
  getRegionsArray: regions,
}) => {
  const selectedName = getSelectedRegion?.name ?? null;
  const handleOnClick = urlId => {
    if (!getRegions[urlId].descriptions) {
      fetchDex(urlId);
    }
    selectRegion(urlId);
  };
  return (
    <ErrorBoundary>
      <div
        data-testid="pokedex-regions"
        className={`${pClassname}__${classname}`}
      >
        <div className={`${pClassname}__${classname}__decor`}>
          <div className={`${pClassname}__${classname}__decor__light`}>
            <div className="trim">
              <div className="bulb red" />
            </div>
          </div>
          <div className={`${pClassname}__${classname}__decor__light`}>
            <div className="trim">
              <div className="bulb yellow" />
            </div>
          </div>
          <div className={`${pClassname}__${classname}__decor__light`}>
            <div className="trim">
              <div className="bulb green" />
            </div>
          </div>
        </div>
        <div className={`${pClassname}__${classname}__list`}>
          {!getRegions && (
            <div className={`${pClassname}__${classname}__list__region noAnim`}>
              <button
                tabIndex={0}
                type={'button'}
                onClick={() => {}}
                onKeyDown={() => {}}
                className={`${pClassname}__${classname}__list__region__button active`}
              >
                national
              </button>
            </div>
          )}
          {!!getRegions &&
            regions.map(({ name, urlId }, index) => (
              <CSSTransition
                key={name}
                timeout={2000}
                in
                unmountOnExit
                appear
                classNames={classNames}
              >
                <div
                  className={`${pClassname}__${classname}__list__region${
                    index === 0 ? ' noAnim' : ''
                  }`}
                >
                  <button
                    key={`${name}-${index}`}
                    tabIndex={0}
                    type={'button'}
                    onClick={() => handleOnClick(urlId)}
                    onKeyDown={() => handleOnClick(urlId)}
                    className={`${pClassname}__${classname}__list__region__button${
                      selectedName === name ? ' active' : ''
                    }`}
                  >
                    {name}
                  </button>
                </div>
              </CSSTransition>
            ))}
        </div>
      </div>
    </ErrorBoundary>
  );
};

Regions.propTypes = {
  pClassname: PropTypes.string.isRequired,
  fetchDex: PropTypes.func.isRequired,
  selectRegion: PropTypes.func.isRequired,
  getRegions: PropTypes.object,
  getRegionsArray: PropTypes.array,
  getSelectedRegion: PropTypes.object,
};

Regions.defaultProps = {
  getRegionsArray: [],
};

export default withPokedex(Regions);

export { Regions };
