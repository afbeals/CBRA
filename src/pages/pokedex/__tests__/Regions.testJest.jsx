// External
import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';

// Local
import { Regions } from '../Regions';

// tests
describe('Regions', () => {
  afterEach(cleanup);
  let fnClick = null;
  beforeEach(() => {
    fnClick = jest.fn();
  });
  it('should match Regions to snapshot', () => {
    const { getByTestId } = render(
      <Regions
        pClassname={'parent'}
        fechDex={fnClick}
        selectRegion={fnClick}
      />,
    );
    const regions = getByTestId('pokedex-regions');
    expect(regions).toMatchSnapshot();
  });

  it('should render the default region', () => {
    const { getByText } = render(
      <Regions
        pClassname={'parent'}
        fechDex={fnClick}
        selectRegion={fnClick}
      />,
    );
    const national = getByText('national');
    expect(national).toBeTruthy();
  });

  it('should render the regions and activate the selected region', () => {
    const regionsArray = [
      { name: 'national', url: '/pokedex/1/', urlId: 1 },
      { name: 'kanto', url: '/pokedex/2/', urlId: 2 },
    ];
    const { getByText, rerender } = render(
      <Regions
        pClassname={'parent'}
        fechDex={fnClick}
        selectRegion={fnClick}
        getRegionsArray={regionsArray}
        getSelectedRegion={{ name: 'national' }}
        getRegions={{
          1: {
            name: 'national',
            descriptions: 'desc',
            urlId: 1,
          },
          2: {
            name: 'kanto',
            descriptions: 'desc',
            urlId: 2,
          },
        }}
      />,
    );
    const nationalRegion = getByText('national');
    const kantoRegion = getByText('kanto');
    expect(nationalRegion).toBeTruthy();
    expect(kantoRegion).toBeTruthy();
    expect(nationalRegion.classList.contains('active')).toBe(true);

    fireEvent.click(kantoRegion);
    rerender(
      <Regions
        pClassname={'parent'}
        fechDex={fnClick}
        selectRegion={fnClick}
        getRegionsArray={regionsArray}
        getSelectedRegion={{ name: 'kanto' }}
        getRegions={{
          1: {
            name: 'national',
            descriptions: 'desc',
            urlId: 1,
          },
          2: {
            name: 'kanto',
            descriptions: 'desc',
            urlId: 2,
          },
        }}
      />,
    );
    expect(kantoRegion.classList.contains('active')).toBe(true);
    expect(nationalRegion.classList.contains('active')).toBe(false);
  });
});
