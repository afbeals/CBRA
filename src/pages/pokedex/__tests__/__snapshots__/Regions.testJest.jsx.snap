// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Regions should match Regions to snapshot 1`] = `
<div
  class="parent__regions"
  data-testid="pokedex-regions"
>
  <div
    class="parent__regions__decor"
  >
    <div
      class="parent__regions__decor__light"
    >
      <div
        class="trim"
      >
        <div
          class="bulb red"
        />
      </div>
    </div>
    <div
      class="parent__regions__decor__light"
    >
      <div
        class="trim"
      >
        <div
          class="bulb yellow"
        />
      </div>
    </div>
    <div
      class="parent__regions__decor__light"
    >
      <div
        class="trim"
      >
        <div
          class="bulb green"
        />
      </div>
    </div>
  </div>
  <div
    class="parent__regions__list"
  >
    <div
      class="parent__regions__list__region noAnim"
    >
      <button
        class="parent__regions__list__region__button active"
        tabindex="0"
        type="button"
      >
        national
      </button>
    </div>
  </div>
</div>
`;
